### PR TITLE
Training with generators

### PR DESCRIPTION
**Pull request by b-biswas**
I have modified the train function so that it can also accept tensorflow.keras.utils.Sequence objects (or data generators) and the training becomes smooth.

### CODE REVIEW

File: debvader/train.py [hunk 1]
```diff
@@ -129,16 +140,17 @@ def kl_metric(y_true, y_pred):
...
+        if not channel_last & (training_data_vae.shape[2] != nb_of_bands):
+            print(
+                "The number of bands in the data does not correspond to the number of filters in the network. Correct this before starting again."
+            )
+            raise ValueError
+        if channel_last & (training_data_vae.shape[-1] != nb_of_bands):
+            print(
+                "The number of bands in the data does not correspond to the number of filters in the network. Correct this before starting again."
+            )
+            raise ValueError
```
Comment (aboucaud) [inline debvader/train.py:153]: Here also

File: debvader/batch_generator.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        return self.num_iterations_per_epoch
+
+    def __getitem__(self, idx):  # WHY idx??
+
+        current_loop_file_name = choice(self.list_of_samples)
+        current_sample = np.load(current_loop_file_name, allow_pickle=True)
+
+        batch = np.random.choice(current_sample, size=self.batch_size, replace=False)
+        x = batch[self.x_col_name]
+        y = batch[self.y_col_name]
```
Comment (aboucaud) [inline debvader/batch_generator.py:48]: Did you use an example to build this batch data generator ?

I am not sure I can follow the logic at the beginning of the `__getitem__`.

Comment (b-biswas) [reply debvader/batch_generator.py:48]: From what I understood, typically a file is storing only one image and __getitem__ usually selects files(or images) whose numbers are given by the parameter `idx`.

But the way my simulations are generated, each file consists of a few hundred images. 
So I randomly select a file from which I want to load data and then extract elements from it.

Comment (b-biswas) [reply debvader/batch_generator.py:48]: I am not sure if this is the right way, I more or less kept to the procedure with which Bastien had trained his old VAE

Comment (aboucaud) [reply debvader/batch_generator.py:48]: So if I understand correctly, you are loading an entire file of ~100 images just to extract a couple of them, right ?

Comment (b-biswas) [reply debvader/batch_generator.py:48]: I am extracting the entire batch of images from the file

**Pull request merged**